zerver/migrations/0372_invalid_characters_in_stream_names.py

```diff
@@ -0,0 +1,66 @@
+import unicodedata
+
+from django.db import connection, migrations
+from django.db.backends.postgresql.schema import DatabaseSchemaEditor
+from django.db.migrations.state import StateApps
+
+
+def fix_stream_names(apps: StateApps, schema_editor: DatabaseSchemaEditor) -> None:
+    Stream = apps.get_model("zerver", "Stream")
+    Realm = apps.get_model("zerver", "Realm")
+
+    total_fixed_count = 0
+    realm_ids = Realm.objects.values_list("id", flat=True)
+    if len(realm_ids) == 0:
+        return
+
+    print("")
+    for realm_id in realm_ids:
+        print(f"Processing realm {realm_id}")
+        realm_stream_dicts = Stream.objects.filter(realm_id=realm_id).values("id", "name")
+        occupied_stream_names = set(stream_dict["name"] for stream_dict in realm_stream_dicts)
+
+        for stream_dict in realm_stream_dicts:
+            stream_name = stream_dict["name"]
+            fixed_stream_name = "".join(
+                [
+                    character
+                    if unicodedata.category(character) not in ["Cc", "Cs"]
+                    else "\N{REPLACEMENT CHARACTER}"
+                    for character in stream_name
```

Comment:
Tweaked this to remove `Cn` from the banned list as discussed, plus added swapping of the characters for `�` - since that's probably better than just removing them.